### PR TITLE
Feedback will now act like TransferFunction in control.

### PR DESCRIPTION
**Pull request by abhiphile**
#### References to other Issues or PRs
Fixes Discussion in https://github.com/sympy/sympy/pull/26228
Fixes Issue https://github.com/sympy/sympy/issues/26161

#### Brief description of what is fixed or changed
1. Feedback in Control Systems will now be used with TransferFunctions easily without affecting the system's overall structure.
2. Added method to find out the numerator and denominator of TransferFunction.
3. Added a method to convert the Feedback system to expression.

#### Release Notes

* physics.control
  * Added support for using Feedback as a TransferFunction.
  * Added methods to get the numerator and denominator of TransferFunction.
  * Added function to get the expression for the Feedback object.

### CODE REVIEW

**Comment by abhiphile** [comment]
@moorepants, Sir, can you please review this PR? I've made sure that everything is correct here.
This can possibly close the issue.

File: sympy/physics/control/lti.py
Comment (moorepants) [inline sympy/physics/control/lti.py:2505]: Can this not simply be `.doit().to_expr()`?

**Comment by moorepants** [comment]
It would be good to have a test showing that you can combine with another transfer function (like putting two a feedback in series with a tf) and show that is works in regular assembly operations.

**Comment by abhiphile** [comment]
> It would be good to have a test showing that you can combine with another transfer function (like putting two a feedback in series with a tf) and show that is works in regular assembly operations.

Sir, Are you talking about this?
` tf3 = TransferFunction(tf1 * fd1, tf2, s) `

**Comment by moorepants** [comment]
> Sir, Are you talking about this?

I was thinking just `Series(Feedback(tf1, tf2), tf3)` then check the num and den are correct and maybe that doit works.

**Comment by abhiphile** [comment]
> I was thinking just `Series(Feedback(tf1, tf2), tf3)` 

I can see that num and den property are not available in `Series` so do I've to implement them inside of `Series`?

File: sympy/physics/control/tests/test_lti.py
Comment (abhiphile) [inline sympy/physics/control/tests/test_lti.py:1032]: I've included the mentioned tests here.

**Comment by abhiphile** [comment]
I can't understand why the tests are failing here.

**Comment by moorepants** [comment]
> I can see that num and den property are not available in Series so do I've to implement them inside of Series?

If making Feedback work as a transfer function means that other things have to be fixed too, then we need to do that. This PR doesn't address the primary issue if you can't use Feedback as a transfer function.

**Comment by moorepants** [comment]
My original example does seem to work with this PR:

```
In [1]: %paste
import sympy as sm
import sympy.physics.control as cn

Ib, Is, c, m, h, g, l2, l1 = sm.symbols('I_b, I_s, c, m, h, g, l2, l1',
                                        real=True, nonnegative=True)
KD, KP, v = sm.symbols('K_D, K_P, v', real=True)
s = sm.symbols('s')

tau1_sq = (Ib + m*h**2)/m/g/h
tau2 = l2/v
tau3 = v/(l1 + l2)
K = v**2/g/(l1 + l2)

Gtheta = cn.TransferFunction(-K*(tau2*s + 1), tau1_sq*s**2 - 1, s)
Gdelta = cn.TransferFunction(1, Is*s**2 + c*s, s)
Gpsi = cn.TransferFunction(1, tau3*s, s)
Dcont = cn.TransferFunction(KD*s, 1, s)
PIcont = cn.TransferFunction(KP, s, s)
Gunity = cn.TransferFunction(1, 1, s)

Ginner = cn.Feedback(Dcont*Gdelta, Gtheta)

Gouter = cn.Feedback(PIcont*Ginner*Gpsi, Gunity)

## -- End pasted text --

In [2]: Gouter
Out[2]: Feedback(Series(TransferFunction(K_P, s, s), Feedback(Series(TransferFunction(K_D*s, 1, s), TransferFunction(1, I_s*s**2 + c*s, s)), TransferFunction(-v**2*(l2*s/v + 1)/(g*(l1 + l2)), -1 + s**2*(I_b + h**2*m)/(g*h*m), s), -1), TransferFunction(1, s*v/(l1 + l2), s)), TransferFunction(1, 1, s), -1)

In [3]: Gouter.num
Out[3]: Series(TransferFunction(K_P, s, s), Feedback(Series(TransferFunction(K_D*s, 1, s), TransferFunction(1, I_s*s**2 + c*s, s)), TransferFunction(-v**2*(l2*s/v + 1)/(g*(l1 + l2)), -1 + s**2*(I_b + h**2*m)/(g*h*m), s), -1), TransferFunction(1, s*v/(l1 + l2), s))

In [4]: Gouter.doit()
Out[4]: TransferFunction(K_D*K_P*g*s**3*v**2*(l1 + l2)*(I_s*s**2 + c*s)**2*(-g*h*m + s**2*(I_b + h**2*m))*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m))), s**2*v*(I_s*s**2 + c*s)*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m)))*(K_D*K_P*g*s*v*(l1 + l2)**2*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m)) + s**2*v*(I_s*s**2 + c*s)*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m))))/(l1 + l2), s)

In [5]: Gouter.doit().to_expr()
Out[5]: (K_D*K_P*g*s**3*v**2*(l1 + l2)*(I_s*s**2 + c*s)**2*(-g*h*m + s**2*(I_b + h**2*m))*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m))))/((s**2*v*(I_s*s**2 + c*s)*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m)))*(K_D*K_P*g*s*v*(l1 + l2)**2*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m)) + s**2*v*(I_s*s**2 + c*s)*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m))))/(l1 + l2)))

In [6]: Gouter.doit().to_expr().simplify()
Out[6]: K_D*K_P*(l1 + l2)**2*(g*h*m - s**2*(I_b + h**2*m))/(K_D*K_P*(l1 + l2)**2*(g*h*m - s**2*(I_b + h**2*m)) + s**2*v*(K_D*h*m*v*(l2*s + v) - (l1 + l2)*(I_s*s + c)*(-g*h*m + s**2*(I_b + h**2*m))))
```

which is nice. It is also a useful test case for the unit tests.

So I guess we can merge this without making Series and Parallel act like a transfer function also. It can be solved separately.

**Comment by abhiphile** [comment]
> tau1_sq = (Ib + m*h**2)/m/g/h
> tau2 = l2/v
> tau3 = v/(l1 + l2)
> K = v**2/g/(l1 + l2)

Thank you for your time , I'm writing a unit test for this also.

File: sympy/physics/control/tests/test_lti.py
Comment (moorepants) [inline sympy/physics/control/tests/test_lti.py:1074]: Equality comparison of long expression is fragile. Best to substract, simiplify and equate to zero.

Comment (moorepants) [reply sympy/physics/control/tests/test_lti.py:1074]: Also, worth checking this is in fact the correct result!

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: I agree Sir, I've reduced the expression

File: sympy/physics/control/tests/test_lti.py
Comment (moorepants) [inline sympy/physics/control/tests/test_lti.py:1069]: This tests the same thing against it itself. It was correct to have the explicitly written expression as before. Just check it by hand that it is correct.

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: I've used this as a reference.

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: And checked the output and I think it is correct.
```
>>> from sympy.physics.control.lti import *
>>> Ib, Is, m, h, l2, l1 = symbols('I_b, I_s, m, h, l2, l1',
...                                             real=True, nonnegative=True)
>>> KD, KP, v = symbols('K_D, K_P, v', real=True)
>>> tau1_sq = (Ib + m * h ** 2) / m / g / h
>>> tau2 = l2 / v
>>> tau3 = v / (l1 + l2)
>>> K = v ** 2 / g / (l1 + l2)
>>> Gtheta = TransferFunction(-K * (tau2 * s + 1), tau1_sq * s ** 2 - 1, s).to_expr()
>>> Gtheta
(-v**2*(l2*s/v + 1)/(g*(l1 + l2)))/(-1 + s**2*(I_b + h**2*m)/(g*h*m))
>>> Gdelta = TransferFunction(1, Is * s ** 2 + c * s, s).to_expr()
>>> from sympy.abc import c
>>> Gdelta = TransferFunction(1, Is * s ** 2 + c * s, s).to_expr()
>>> Gdelta
1/(I_s*s**2 + c*s)
>>> Gpsi = TransferFunction(1, tau3 * s, s).to_expr()
>>> Dcont = TransferFunction(KD * s, 1, s).to_expr()
>>> Dcont
(K_D*s)/1
>>> PIcont = TransferFunction(KP, s, s).to_expr()
>>> Gunity = TransferFunction(1, 1, s).to_expr()
>>> expnum = PIcont * Dcont * Gdelta / (s * tau3 * (1 + Dcont * Gdelta * Gtheta))
>>> expnum
K_D*K_P*(l1 + l2)/(s*v*(I_s*s**2 + c*s)*(-K_D*s*v**2*(l2*s/v + 1)/(g*(-1 + s**2*(I_b + h**2*m)/(g*h*m))*(l1 + l2)*(I_s*s**2 + c*s)) + 1))
>>> expden = 1 + PIcont*Dcont*Gdelta*Gpsi*Gunity/(1 + Dcont*Gdelta*Gtheta)
>>> expden
K_D*K_P*(l1 + l2)/(s*v*(I_s*s**2 + c*s)*(-K_D*s*v**2*(l2*s/v + 1)/(g*(-1 + s**2*(I_b + h**2*m)/(g*h*m))*(l1 + l2)*(I_s*s**2 + c*s)) + 1)) + 1
>>> exp = expnum/expden

>>> exp
K_D*K_P*(l1 + l2)/(s*v*(I_s*s**2 + c*s)*(K_D*K_P*(l1 + l2)/(s*v*(I_s*s**2 + c*s)*(-K_D*s*v**2*(l2*s/v + 1)/(g*(-1 + s**2*(I_b + h**2*m)/(g*h*m))*(l1 + l2)*(I_s*s**2 + c*s)) + 1)) + 1)*(-K_D*s*v**2*(l2*s/v + 1)/(g*(-1 + s**2*(I_b + h**2*m)/(g*h*m))*(l1 + l2)*(I_s*s**2 + c*s)) + 1))

>>> tf = TransferFunction(KD*KP*g*s**3*v**2*(l1+l2)*(Is*s**2+c*s)**2*(-g*h*m+s**2*(Ib+h**2*m))*(-KD*g*h*m*s*v**2*(l2*s+v)+g*v*(l1+l2)*(Is*s**2+c*s)*(-g*h*m+s**2*(Ib+h**2*m))), s**2*v*(Is*s**2+c*s)*(-KD*g*h*m*s*v**2*(l2*s+v)+g*v*(l1+l2)*(Is*s**2+c*s)*(-g*h*m+s**2*(Ib+h**2*m)))*(KD*KP*g*s*v*(l1+l2)**2*(Is*s**2+c*s)*(-g*h*m+s**2*(Ib+h**2*m))+s**2*v*(Is*s**2+c*s)*(-KD*g*h*m*s*v**2*(l2*s+v)+g*v*(l1+l2)*(Is*s**2+c*s)*(-g*h*m+s**2*(Ib+h**2*m))))/(l1+l2), s)

>>> exp1 = tf.to_expr()

>>> exp1 - exp
-K_D*K_P*(l1 + l2)/(s*v*(I_s*s**2 + c*s)*(K_D*K_P*(l1 + l2)/(s*v*(I_s*s**2 + c*s)*(-K_D*s*v**2*(l2*s/v + 1)/(g*(-1 + s**2*(I_b + h**2*m)/(g*h*m))*(l1 + l2)*(I_s*s**2 + c*s)) + 1)) + 1)*(-K_D*s*v**2*(l2*s/v + 1)/(g*(-1 + s**2*(I_b + h**2*m)/(g*h*m))*(l1 + l2)*(I_s*s**2 + c*s)) + 1)) + (K_D*K_P*g*s**3*v**2*(l1 + l2)*(I_s*s**2 + c*s)**2*(-g*h*m + s**2*(I_b + h**2*m))*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m))))/((s**2*v*(I_s*s**2 + c*s)*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m)))*(K_D*K_P*g*s*v*(l1 + l2)**2*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m)) + s**2*v*(I_s*s**2 + c*s)*(-K_D*g*h*m*s*v**2*(l2*s + v) + g*v*(l1 + l2)*(I_s*s**2 + c*s)*(-g*h*m + s**2*(I_b + h**2*m))))/(l1 + l2)))


>>> simplify(exp1-exp)
0
```

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: I have checked everything properly, and I think the results are correct.

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: @moorepants Can you please check this !

File: sympy/physics/control/lti.py
Comment (moorepants) [inline sympy/physics/control/lti.py:2406]:
```suggestion
        Returns the numerator of the closed loop feedback system.
```

File: sympy/physics/control/lti.py
Comment (moorepants) [inline sympy/physics/control/lti.py:2413]:
```suggestion
        Returns the denominator of the closed loop feedback model.
```

File: sympy/physics/control/lti.py
Comment (moorepants) [inline sympy/physics/control/lti.py:2501]: Actually, do you need this defined at all? Doesn't TransferFunction do the same thing and this behavior is inherited?

Comment (abhiphile) [reply sympy/physics/control/lti.py:2501]: No Sir, It was giving an error when I've not defined it.

File: sympy/physics/control/tests/test_lti.py
Comment (moorepants) [reply sympy/physics/control/tests/test_lti.py:1074]: Sorry I think you keep misunderstanding me.

1. Yes it is correct to write the full known correct result in the test, like you have done. So keep that.
2. You should do `(known - calculated).simplify == 0` for the test to ensure a robust symbolic test.

File: sympy/physics/control/lti.py
Comment (moorepants) [reply sympy/physics/control/lti.py:2501]: to_expr() does not call doit() in TF:

https://github.com/sympy/sympy/blob/master/sympy/physics/control/lti.py#L1273

So this is fine, but you need a docstring.

File: sympy/physics/control/tests/test_lti.py
Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py]: Thank you, sir. I apologize for any mistakes in my recent contributions to the control. As I recently started contributing, I'm still learning new things, and I appreciate your guidance in helping me improve.

Comment (moorepants) [reply sympy/physics/control/tests/test_lti.py]: There is no need to apologize, nor to call me sir. Your mistakes are fine and you are very close to having this merged.

File: sympy/physics/control/lti.py
Comment (abhiphile) [reply sympy/physics/control/lti.py:2519]: I've written the docstring in the recent commit.

**Review by faze-geek — APPROVED** [review]
Great reviews and work here @abhiphile. I'll wait for Jason to merge this when he feels appropriate. Thanks !

File: sympy/physics/control/tests/test_lti.py
Comment (moorepants) [reply sympy/physics/control/tests/test_lti.py:1069]: Sorry but I don't think this test is a good one. You had the explicit expression written out and tested against it at one point and that is the correct way to do it. With the test you have, if we break something in the logic then both sides of the equality will break and the equality can still hold. Test against the final expression that we know is mathematically correct.

**Comment by moorepants** [comment]
LGTM, nice work!

**Comment by oscarbenjamin** [comment]
> Failing test here doesn't seem to be unrelated.

That particular pypy test just fails randomly since some recent pypy release (a few months ago?). I can't reproduce the failure outside of CI so I haven't opened a PyPy issue about it. The problem is something to do with pytest's assertion rewriting which presumably messes with PyPy's AST.

https://github.com/pypy/pypy/issues/3959
https://github.com/pytest-dev/pytest/issues/11168

**Comment by faze-geek** [comment]
Sorry, mentioned "doesn't" and "unrelated" together by mistake. Did not mean to cause a confusion 😄

**Pull request merged**
